servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/MulticastSingle.java

```diff
@@ -0,0 +1,371 @@
+         * </ul>
+         */
+        @Nullable
+        private volatile Object state = INIT;
+
+        private ConcurrentOnSubscribeSubscriber(final Subscriber<? super T> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onSubscribe(final Cancellable cancellable) {
+            for (;;) {
+                final Object currState = state;
+                if (currState == INIT && stateUpdater.compareAndSet(this, INIT, INVOKING_ON_SUBSCRIBE)) {
+                    try {
+                        delegate.onSubscribe(cancellable);
+                    } finally {
+                        if (!stateUpdater.compareAndSet(this, INVOKING_ON_SUBSCRIBE, WAITING_FOR_TERMINAL)) {
+                            sendTerminal(state); // re-read state which has been changed by another thread.
+                        }
+                    }
+                    break;
+                } else if (currState == WAITING_FOR_TERMINAL || currState == INVOKING_ON_SUBSCRIBE ||
+                        currState == TERMINATED) {
+                    duplicateOnSubscribe(cancellable);
+                    break;
+                } else if (currState != INIT) {
+                    delayedOnSubscribe(currState);
+                    break;
+                }
+            }
+        }
+
+        private void duplicateOnSubscribe(Cancellable cancellable) {
+            try {
+                cancellable.cancel(); // duplicate onSubscribe not supported
+            } finally {
+                logDuplicateTerminal(this);
+            }
+        }
+
+        private void delayedOnSubscribe(@Nullable Object currState) {
+            try {
+                delegate.onSubscribe(IGNORE_CANCEL);
+            } finally {
+                sendTerminal(currState);
+            }
+        }
+
+        private void sendTerminal(@Nullable Object currState) {
+            state = TERMINATED;
+            if (currState instanceof TerminalNotification) {
+                Throwable cause = ((TerminalNotification) currState).cause();
+                assert cause != null;
+                delegate.onError(cause);
+            } else {
+                @SuppressWarnings("unchecked")
+                final T t = (T) currState;
+                delegate.onSuccess(t);
+            }
+        }
+
+        @Override
+        public void onSuccess(@Nullable final T result) {
+            for (;;) {
+                final Object currState = state;
+                if (currState == WAITING_FOR_TERMINAL) {
+                    state = TERMINATED; // onSubscribe already invoked, no need for atomic as no concurrency allowed.
+                    delegate.onSuccess(result);
+                    break;
+                } else if (currState == INIT) {
+                    if (stateUpdater.compareAndSet(this, INIT, TERMINATED)) {
+                        deliverSuccessFromSource(delegate, result);
```

Comment:
Let's consider a case when Thread-onSubscribe and Thread-onSuccess both enter `currState == INIT`:
1. Thread-onSuccess wins cas -> swaps the state to `TERMINATED` -> invokes `onSubscribe` -> delivers terminal
2. Thread-onSubscribe goes to the second for loop iteration -> sees `currState == TERMINATED` -> invokes `duplicateOnSubscribe`

We should be able to distinguish this situation from the real duplicate onSubscribe.

Reply:
good catch, I will adjust. we don't need to differentiate between duplicate onSubscribe and race between terminal/onSubscribe (`logDuplicateTerminal` is best effort) but we just need to make sure we terminate form the loop.